cpp/include/ray/api.h

```diff
@@ -49,7 +49,7 @@ class Ray {
   /// \param[in] object The object reference which should be returned.
   /// \return shared pointer of the result.
   template <typename T>
-  static std::shared_ptr<T> Get(const ObjectRef<T> &object);
+  static T Get(const ObjectRef<T> &object);
```

Comment:
I don't think this is a good change here for some reasons:

1. the return value here is returned by copy, even if there might be some return-value-optimizations from complier.
2. returning shared_ptrs  sounds cpp-native enough.

Reply:
In c++11, the return value will be always moved, no copy here.

Reply:
What about the unmovable objects?

Reply:
If the user forbidden the move constructor and move assignable constructor of T, the return T will compile error, no side effects。
So the user should return a smart pointer of the object, when it is uncopyable or unmovable.

Reply:
If the object is cached in core worker memory, it will be copied.

Reply:
The c++11 RVO make sure there is no copy, if cached the objects, we can move it into cache to avoid copy, another way is use smart pointer as the return type in user function.

In other word, we should not restrict the return type only as shared_ptr, some times it maybe an object, sometimes it maybe a smart pointer, no restriction here, it's up to the user, and the Get interface will be more natural and easy to use.

Reply:
> If the object is cached in core worker memory, it will be copied.

I think this is OK. We already did this in Python and Java. It's the user's responsibility to avoid getting the same object twice.

Reply:
> What about the unmovable objects?

I guess we should treat unmovable objects as un-serializable, right? If a class/struct only contains data, there's no reason to delete the move constructor.

Reply:
> In c++11, the return value will be always moved, no copy here.

I think it's not in C++ 11 standard, but C++ 17. Am I missing something here? Reference: https://jonasdevlieghere.com/guaranteed-copy-elision/

Reply:
Make a clarification, RVO is recommended not required in c++11, actually most compilers have support it in c++11.